### PR TITLE
8292275: javac does not emit SYNTHETIC and MANDATED flags for parameters by default

### PR DESCRIPTION
**Pull request by SirYwell**
With this change, javac emits the MethodParameters attribute in cases where the JLS requires the information about synthetic and mandated parameters to be stored (see issue).
Parameter names are *not* emitted unless the `-parameter` flag is set.

The relevant changes are in `ClassWriter`, where we go through the params to see if we need the attribute if the `-parameter` flag is not set (if it is set, both names and flags will be emitted).
For records, the mandated flag wasn't set at all, this is solved by the one line fix in `JavacParser`.

The changes to `CreateSymbols` and `ClassReader` are needed as they weren't able to deal with missing names in the attribute.
I also had to update some tests as they got a new constant pool entry.

Only the mandated flag is covered by tests at the moment, as the occurrences are well-specified in the JLS.
Please let me know if you want tests for specific appearances of synthetic parameters.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8292467](https://bugs.openjdk.org/browse/JDK-8292467) to be approved

### Issues
 * [JDK-8292275](https://bugs.openjdk.org/browse/JDK-8292275): javac does not emit SYNTHETIC and MANDATED flags for parameters by default
 * [JDK-8292467](https://bugs.openjdk.org/browse/JDK-8292467): javac does not emit SYNTHETIC and MANDATED flags for parameters by default (**CSR**)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to [5ac3a214](https://git.openjdk.org/jdk/pull/9862/files/5ac3a214be729cd34e37d2ad07ad504328d4c1b8)
 * [Julian Waters](https://openjdk.org/census#jwaters) (@TheShermanTanker - Committer)

### Contributors
 * Chen Liang `<liach@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/9862/head:pull/9862` \
`$ git checkout pull/9862`

Update a local copy of the PR: \
`$ git checkout pull/9862` \
`$ git pull https://git.openjdk.org/jdk.git pull/9862/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9862`

View PR using the GUI difftool: \
`$ git pr show -t 9862`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9862.diff">https://git.openjdk.org/jdk/pull/9862.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/9862#issuecomment-1213484311)

### CODE REVIEW

File: test/langtools/tools/javac/RequiredParameterFlags/ImplicitParameters.java
Comment (jddarcy) [inline test/langtools/tools/javac/RequiredParameterFlags/ImplicitParameters.java:52]: In addition to this test, an annotation processing-based test could also be attempted here.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java
Comment (jddarcy) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java:4023]: Should the use of MANDATED be conditional on the target class file version?

Comment (jddarcy) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java:4023]: I advise also writing a core reflection test that uses access flags (java.lang.reflect.AccessFlag).

Comment (SirYwell) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java:4023]: > Should the use of MANDATED be conditional on the target class file version?

I don't think this is needed. The JLS mentions the parameters of canonical record constructors in Java 14 already. Previous versions don't have records and shouldn't enter that method at all.

> I advise also writing a core reflection test that uses access flags (java.lang.reflect.AccessFlag).

Will do. I assume the `jdk/java/lang/reflect/AccessFlag` folder is a good place for that?

Comment (jddarcy) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java:4023]: > > Should the use of MANDATED be conditional on the target class file version?
> 
> I don't think this is needed. The JLS mentions the parameters of canonical record constructors in Java 14 already. Previous versions don't have records and shouldn't enter that method at all.

Agreed; I double-checked and the ACC_MANDATED flag was defined in that location since before JDK 14's format.

> 
> > I advise also writing a core reflection test that uses access flags (java.lang.reflect.AccessFlag).
> 
> Will do. I assume the `jdk/java/lang/reflect/AccessFlag` folder is a good place for that?

Yes; that an appropriate  home for such tests.

File: test/langtools/tools/javac/RequiredParameterFlags/ImplicitParameters.java
Comment (SirYwell) [reply test/langtools/tools/javac/RequiredParameterFlags/ImplicitParameters.java:52]: Could you elaborate on that? Should I add a processor to the JavacTask in this test or write a separate one? (Maybe there is a similar test I can use as reference?)
Thank you.

**Comment by vicente-romero-oracle** [comment]
CSR needed I guess

File: test/langtools/tools/javac/RequiredParameterFlags/ImplicitParameters.java
Comment (jddarcy) [reply test/langtools/tools/javac/RequiredParameterFlags/ImplicitParameters.java:52]: > Could you elaborate on that? Should I add a processor to the JavacTask in this test or write a separate one? (Maybe there is a similar test I can use as reference?) Thank you.

There are tests that can be used as a model around

test/langtools/tools/javac/processing/model/util/elements 

HTH

Comment (SirYwell) [reply test/langtools/tools/javac/RequiredParameterFlags/ImplicitParameters.java:52]: Sorry, it looks like there is no way to access the synthetic/mandated information using the annotation processing API. Am I missing something?

**Comment by SirYwell** [comment]
> CSR needed I guess

I don't have access to the JBS. Could you help me with this?

**Comment by TheShermanTanker** [comment]
> > CSR needed I guess
> 
> I don't have access to the JBS. Could you help me with this?

No problem!

File: test/langtools/tools/javac/RequiredParameterFlags/ImplicitParameters.java
Comment (jddarcy) [reply test/langtools/tools/javac/RequiredParameterFlags/ImplicitParameters.java:52]: > Sorry, it looks like there is no way to access the synthetic/mandated information using the annotation processing API. Am I missing something?

For reasons of providing a minimal core language model that could be reused, less essential functionality is split in the language model API to be provided by an implementation of javax.lang.model.util.Elements.

So if you subclass JavacTestingAbstractProcessor files that test directory, you could use

eltUtils.getOrigin(elementForParameter)

to see if an element is marked as javax.lang.model.util.Enum Elements.Origin.{MANDATED, SYNTHETIC}.

Many of the annotation processing tests use a pattern of using a runtime-retention annotation to encode what the expected answer is. In this case, I'd recommend a method/constructor annotation to hold the expected origin of the parameters.

(The annotation processing API does filter out some SYNTHETIC constructs, but I don't recall if that is done at a parameter level.)

HTH

Comment (SirYwell) [reply test/langtools/tools/javac/RequiredParameterFlags/ImplicitParameters.java:52]: Thanks, I started working on a test, but the results are somewhat inconsistent (no implicit parameter is present for inner class ctors, no synthetic parameters for enum ctors, parameter of valueOf is MANDATED, implicit parameters in compact record ctor are MANDATED).
I'm not sure how we want to proceed with that or if there are known issues about that, but it might be better to tackle that separately. What do you think?

Comment (jddarcy) [reply test/langtools/tools/javac/RequiredParameterFlags/ImplicitParameters.java:52]: > Thanks, I started working on a test, but the results are somewhat inconsistent (no implicit parameter is present for inner class ctors, no synthetic parameters for enum ctors, parameter of valueOf is MANDATED, implicit parameters in compact record ctor are MANDATED). I'm not sure how we want to proceed with that or if there are known issues about that, but it might be better to tackle that separately. What do you think?

At least some of that behavior is expected/not surprising given the current state of javac's implementation. I linked the existing bug related bug

    JDK-8250919: Mark compiler-generated elements as mandated

to JDK-8292275 in JBS. When possible, arguably javac should present the "source view" of a construct, so it would omit, say, the synthetic parameters javac chooses to add to enum constructors.

I suggest including a test for the mandated structures that "work" and the test could be expanded as that property is tracked and used more uniformly.

Thanks and HTH

File: test/langtools/tools/javac/processing/model/util/elements/MethodParameters/ClassContainer.java
Comment (jddarcy) [inline test/langtools/tools/javac/processing/model/util/elements/MethodParameters/ClassContainer.java:1]: I recommend folding the enum being processed into the source of ImplicitParametersProcessor.java. The jtreg tags can then first compile the source of the processor and then secondly run the processor on itself. This technique is commonly used in annotation processing tests.

File: test/langtools/tools/javac/processing/model/util/elements/MethodParameters/ImplicitParametersProcessor.java
Comment (jddarcy) [inline test/langtools/tools/javac/processing/model/util/elements/MethodParameters/ImplicitParametersProcessor.java:33]: Good start on the annotation processor. From personal experience writing such tests and then looking at them months later, I recommend adding some kind of brief comment describing the intended semantics of the test, e.g.

"Verify enum classes and record classes have expected mandated executables."

To make the test more extensible, the expected information could be encoded in annotations; e.g. something like

@interface ExpectedInfo {
    String methodName();
    Elements.Origin origin();
}

where the record and enum were annotated in turn:

@ExpectedInfo(name="<init>", origin=MANDATED)
MyEnum {..}

etc.

**Comment by SirYwell** [comment]
Hi, I addressed your comments. The annotation is definitely a cleaner solution, thank you for that suggestion.

**Comment by SirYwell** [comment]
Hi, is there anything else that needs to be done from my side?

**Comment by SirYwell** [comment]
Hi, I merged master recently and updated the copyright year. Is there anything else I can do to get the CSR approved and my changes integrated?

**Comment by SirYwell** [comment]
Still waiting for any kind of feedback.

**Comment by jddarcy** [comment]
@vicente-romero-oracle , can you review the CSR for this? Thanks.

**Comment by vicente-romero-oracle** [comment]
> @vicente-romero-oracle , can you review the CSR for this? Thanks.

sure, taking a look now

**Comment by SirYwell** [comment]
@vicente-romero-oracle thank you for the feedback. As I don't have a JBS account, I can't incorporate it myself. @TheShermanTanker could you do that for me as you created the CSR? Thanks.

**Comment by vicente-romero-oracle** [comment]
> @vicente-romero-oracle thank you for the feedback. As I don't have a JBS account, I can't incorporate it myself. @TheShermanTanker could you do that for me as you created the CSR? Thanks.

I will incorporate the changes to the CSR, please let me know if you have any suggestions

**Comment by SirYwell** [comment]
> I will incorporate the changes to the CSR, please let me know if you have any suggestions

THanks again. I don't have any more suggestions.

**Comment by liach** [comment]
Just checking on this. the CSR [has a response](https://bugs.openjdk.org/browse/JDK-8292467?focusedCommentId=14562384&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14562384):

> Please refine the discussion in "Solution" to explicitly state which --release/-target versions are intended to be impacted by the change. Presumably, at a minimum the synthetic and mandated information would only be written to class file formats where those bits are defined.
>
> A more conservative form of this change would be to only change the behavior for --release/-target 21. However, given the low expected impact of making the change, I'd expect it would be fine to make the change in behavior for all class file versions where that information is supported.

Our current implementation applies the change of behavior to all `--release/-target` versions. And as far as I see, `MethodParameters` attribute has been properly supported by the Reflection API to filter out synthetic and mandated parameters since 1c54a00f752a54a7a68b09d3df5d3a6081f4fcc8 (Java 8, when `MethodParameters` is introduced). So I suggest to move along by specifying that this change applies to all release/target versions, which will fix the reflection issue on all versions.

It ultimately depends on @SirYwell to decide what to do here. I can help you update the CSR's Solution Section as you desire, as I am eager to see this bug fixed on javac's end as well.

File: test/langtools/tools/javac/RequiredParameterFlags/ImplicitParameters.java
Comment (liach) [inline test/langtools/tools/javac/RequiredParameterFlags/ImplicitParameters.java:83]: We really need another local class example, such as
```java
{
    int apple = Integer.parseInt("apple");
    Runnable run = new Runnable() {
        public void run() {
            System.out.println(apple);
        }
    }
}
```

**Comment by liach** [comment]
This appear related to https://bugs.openjdk.org/browse/JDK-8213329 as well, for this patch makes javac emit parameter attributes reliably.

**Comment by SirYwell** [comment]
Sorry for the late response.

> Our current implementation applies the change of behavior to all `--release/-target` versions. And as far as I see, `MethodParameters` attribute has been properly supported by the Reflection API to filter out synthetic and mandated parameters since [1c54a00](https://github.com/openjdk/jdk/commit/1c54a00f752a54a7a68b09d3df5d3a6081f4fcc8) (Java 8, when `MethodParameters` is introduced). So I suggest to move along by specifying that this change applies to all release/target versions, which will fix the reflection issue on all versions.

I agree with this, I'd be happy if you could update the CSR accordingly.

I will try to address your review comment as soon as possible.

**Comment by liach** [comment]
> I agree with this, I'd be happy if you could update the CSR accordingly.

I've updated and finalized the CSR for you.

> I will try to address your review comment as soon as possible.

I've opened SirYwell/jdk#1 with 2 commits to address my review concerns about Tests. Feel free to check it out, though the PR will be hard to navigate until you merge jdk's master branch into your branch.

**Comment by SirYwell** [comment]
Test failures are caused by the Classfile API. I opened a separate PR to fix it: https://github.com/openjdk/jdk/pull/13167

Could someone create an issue on the bug tracker for me? Thanks.

**Comment by TheShermanTanker** [comment]
@SirYwell On it!

**Comment by liach** [comment]
@vicente-romero-oracle Would you take a look at this patch since you've reviewed the associated CSR already? This is a helpful fix that can allow core reflection to map annotations and signatures more accurately with the newly available MethodParameters information.

File: make/langtools/src/classes/build/tools/symbolgenerator/CreateSymbols.java
Comment (vicente-romero-oracle) [inline make/langtools/src/classes/build/tools/symbolgenerator/CreateSymbols.java:2387]: could use ternary operator here

File: make/langtools/src/classes/build/tools/symbolgenerator/CreateSymbols.java
Comment (vicente-romero-oracle) [inline make/langtools/src/classes/build/tools/symbolgenerator/CreateSymbols.java:2388]: could use ternary operator here

Comment (liach) [reply make/langtools/src/classes/build/tools/symbolgenerator/CreateSymbols.java:2388]: `name_index` could always have been 0 to indicate an absent parameter name: https://docs.oracle.com/javase/specs/jvms/se20/html/jvms-4.html#jvms-4.7.24

Comment (vicente-romero-oracle) [reply make/langtools/src/classes/build/tools/symbolgenerator/CreateSymbols.java:2388]: I see, then this was a bug in the previous code, still I suggest using the ternary operator here

**Comment by liach** [comment]
> this test: jdk/classfile/CorpusTest.java is consistently failing with this fix applied

This test reads the jdk image for its class files. Without #13167's fix to Classfile API, it will fail on the jdk image with MethodParameters with absent parameter names generated by this javac update. Can you test again with that patch integrated as well?

**Comment by vicente-romero-oracle** [comment]
> /CorpusTest.java

sure

**Comment by vicente-romero-oracle** [comment]
> 

sure will test with both fixes applied

File: make/langtools/src/classes/build/tools/symbolgenerator/CreateSymbols.java
Comment (liach) [inline make/langtools/src/classes/build/tools/symbolgenerator/CreateSymbols.java:2390]:
```suggestion
                    String name = e.name_index == 0 ? null
                            : cf.constant_pool.getUTF8Value(e.name_index);
```
assuming a continuation indent of 8.

**Comment by vicente-romero-oracle** [comment]
I have approved: https://github.com/openjdk/jdk/pull/13167

**Review by vicente-romero-oracle — APPROVED** [review]
looks good

**Comment by vicente-romero-oracle** [comment]
please remember to wait for https://github.com/openjdk/jdk/pull/13167 to be integrated before integrating this one

**Review by TheShermanTanker — APPROVED** [review]

**Comment by SirYwell** [comment]
Switched to the ternary operator as suggested, thanks for your review @vicente-romero-oracle.

**Review by TheShermanTanker — APPROVED** [review]

**Comment by TheShermanTanker** [comment]
Windows Failures are due to a longstanding (and annoying) build system bug that's been around for quite a while by this point, and aren't related to this PR in any way

Shall I do the honours?

**Comment by liach** [comment]
> Windows Failures are due to a longstanding (and annoying) build system bug that's been around for quite a while by this point, and aren't related to this PR in any way
> 
> Shall I do the honours?

If it's the vscode toolchain one, go for it. It fixes itself on github from time to time, and definitely doesn't break anything when built by openjdk or adoptium.

**Comment by TheShermanTanker** [comment]
Alright, here goes

**Comment by liach** [comment]
Nice! One of the most troubling java reflection bugs will be no more:

```java
public enum TestEnum {
    ;
    TestEnum(Optional<String> genericParameter) {}
}
```
with this, `TestEnum.class.getDeclaredConstructors()[0].getParameters()[2].getParameterizedType()` will return `java.util.Optional<java.lang.String>` than `class java.util.Optional`, which is why I've been eager for this patch.

**Comment by TheShermanTanker** [comment]
Haha, glad to have been of help

Now, if only someone would also take a look at my javac patch for the JNI header generator... :(

**Comment by vicente-romero-oracle** [comment]
good to see this one done, thanks to all!

**Comment by vicente-romero-oracle** [comment]
> Haha, glad to have been of help
> 
> Now, if only someone would also take a look at my javac patch for the JNI header generator... :(

what's the PR for that one?

**Comment by TheShermanTanker** [comment]
> > Haha, glad to have been of help
> > Now, if only someone would also take a look at my javac patch for the JNI header generator... :(
> 
> what's the PR for that one?

Ah, sorry for the trouble, I was mildly joking for that one, but the corresponding PR is https://github.com/openjdk/jdk/pull/13457

**Comment by vicente-romero-oracle** [comment]
> > > Haha, glad to have been of help
> > > Now, if only someone would also take a look at my javac patch for the JNI header generator... :(
> > 
> > 
> > what's the PR for that one?
> 
> Ah, sorry for the trouble, I was mildly joking for that one, but the corresponding PR is #13457

I see, it seems like it is already being looked at

**Comment by TheShermanTanker** [comment]
It was yeah, Jan hasn't come back to reviewing it in quite a while unfortunately :/

**Pull request closed without merging**
